### PR TITLE
Improve the height calculation of the ConsoleLogStream component

### LINKED ISSUE #370: ConsoleLogStream Height is Inconsistent or Wrong
### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  
  ```text
Jenkins: 2.440
OS: Mac OS X - 13.5.1
Java: 11.0.14.1 - Eclipse Adoptium (OpenJDK 64-Bit Server VM)
---
apache-httpcomponents-client-4-api:4.5.14-208.v438351942757
bouncycastle-api:2.30.1.77-225.v26ea_c9455fd9
branch-api:2.1144.v1425d1c3d5a_7
caffeine-api:3.1.8-133.v17b_1ff2e0599
cloudbees-folder:6.858.v898218f3609d
commons-lang3-api:3.13.0-62.v7d18e55f51e2
credentials:1311.vcf0a_900b_37c2
credentials-binding:657.v2b_19db_7d6e6d
display-url-api:2.200.vb_9327d658781
durable-task:543.v262f6a_803410
git:5.2.1
git-client:4.6.0
github:1.37.3.1
github-api:1.318-461.v7a_c09c9fa_d63
github-branch-source:1772.va_69eda_d018d4
gson-api:2.10.1-15.v0d99f670e0a_7
instance-identity:185.v303dc7c645f9
ionicons-api:56.v1b_1c8c49374e
jackson2-api:2.16.1-373.ve709c6871598
jakarta-activation-api:2.0.1-3
jakarta-mail-api:2.0.1-3
javax-activation-api:1.2.0-6
javax-mail-api:1.6.2-9
jaxb:2.3.9-1
jjwt-api:0.11.5-77.v646c772fddb_0
json-path-api:2.8.0-21.v8b_7dc8b_1037b_
mailer:463.vedf8358e006b_
metrics:4.2.18-442.v02e107157925
mina-sshd-api-common:2.12.0-90.v9f7fb_9fa_3d3b_
mina-sshd-api-core:2.12.0-90.v9f7fb_9fa_3d3b_
okhttp-api:4.11.0-157.v6852a_a_fa_ec11
pipeline-build-step:540.vb_e8849e1a_b_d8
pipeline-graph-analysis:202.va_d268e64deb_3
pipeline-graph-view:232.vc7ca_8d934725
pipeline-groovy-lib:700.v0e341fa_57d53
pipeline-input-step:477.v339683a_8d55e
pipeline-model-api:2.2168.vf921b_4e72c73
pipeline-model-definition:2.2168.vf921b_4e72c73
pipeline-model-extensions:2.2168.vf921b_4e72c73
pipeline-stage-step:305.ve96d0205c1c6
pipeline-stage-tags-metadata:2.2168.vf921b_4e72c73
plain-credentials:143.v1b_df8b_d3b_e48
scm-api:683.vb_16722fb_b_80b_
script-security:1313.v7a_6067dc7087
snakeyaml-api:2.2-111.vc6598e30cc65
ssh-credentials:308.ve4497b_ccd8f4
structs:337.v1b_04ea_4df7c8
token-macro:400.v35420b_922dcb_
trilead-api:2.133.vfb_8a_7b_9c5dd1
variant:60.v7290fc0eb_b_cd
workflow-api:1283.v99c10937efcb_
workflow-basic-steps:1042.ve7b_140c4a_e0c
workflow-cps:3837.v305192405b_c0
workflow-durable-task-step:1313.vcb_970b_d2a_fb_3
workflow-job:1385.vb_58b_86ea_fff1
workflow-multibranch:773.vc4fe1378f1d5
workflow-scm-step:415.v434365564324
workflow-step-api:657.v03b_e8115821b_
workflow-support:865.v43e78cc44e0d
  ```

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

OS: Mac OS X - 13.5.1

### Reproduction steps

1. Create a simple pipeline with varying log lengths:
```
stage("A") {
    sleep(30)
    def message = "Hello World\n"
    echo(message*1)
    echo(message*2)
    echo(message*5)
    echo(message*10)
    echo(message*20)
    echo(message*40)
    echo(message*70)
    echo(message*110)
    echo(message*160)
    echo(message*200)
    echo(message*201)
}
```
2. Run the pipeline job and observe issues with the heights of each [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) component.

### Expected Results

Each [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) component is created with the correct size to just contain its contents.

### Actual Results

There is increasing white space after the logs for each [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) instance

image

If there are between ~33 and 199 line logs then the height of the [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) will be proportionally large, however after 200 lines the component will be reduced to `60vh`. This is an inconsistent experience for the user.

The console modal does not expand the  [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) component to use its full height.

image

If there are only a few log lines for a running step (`sleep` or `build` for example), the [`ConsoleLogStream`](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) component won't be tall enough to include both the log lines and the spinner icon.

Screen Recording 2024-03-21 at 10 37 52 AM

### Anything else?

_No response_

### Are you interested in contributing a fix?

Yes.

### PR DESCRIPTION
**Pull request by stuartrowe**
Updated the minHeight calculation for the [ConsoleLogStream](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) component so it's more appropriately sized for its content.
 
Screenshot 2024-03-21 at 11 01 13 AM
Screenshot 2024-03-21 at 11 02 12 AM

Also updated the calculation to account for the spinner icon for in progress runs. 
Screen Recording 2024-03-21 at 11 02 32 AM

Increased the max height scale of the [ConsoleLogStream](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx) when used by the [ConsoleLogModal](https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogModal.tsx) so it uses more of the available space.

Screenshot 2024-03-21 at 11 23 08 AM

Fixes #370.

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogCard.tsx [hunk 1]
```diff
@@ -276,6 +276,7 @@ export class ConsoleLogCard extends React.Component<
           logBuffer={this.props.stepBuffer}
           handleMoreConsoleClick={this.props.handleMoreConsoleClick}
           step={this.props.step}
+          maxHeighScale={0.85}
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogCard.tsx:279]:
```suggestion
          maxHeightScale={0.85}
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogCard.tsx [hunk 2]
```diff
@@ -295,6 +296,7 @@ export class ConsoleLogCard extends React.Component<
                 logBuffer={this.props.stepBuffer}
                 handleMoreConsoleClick={this.props.handleMoreConsoleClick}
                 step={this.props.step}
+                maxHeighScale={0.5}
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogCard.tsx:299]:
```suggestion
                maxHeightScale={0.5}
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogModal.tsx [hunk 3]
```diff
@@ -7,6 +7,7 @@ export interface ConsoleLogModelProps {
   logBuffer: StepLogBufferInfo;
   handleMoreConsoleClick: (nodeId: string, startByte: number) => void;
   step: StepInfo;
+  maxHeighScale: number;
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogModal.tsx:10]:
```suggestion
  maxHeightScale: number;
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx [hunk 4]
```diff
@@ -9,6 +9,7 @@ export interface ConsoleLogStreamProps {
   logBuffer: StepLogBufferInfo;
   handleMoreConsoleClick: (nodeId: string, startByte: number) => void;
   step: StepInfo;
+  maxHeighScale: number;
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx:12]:
```suggestion
  maxHeightScale: number;
```

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx [hunk 5]
```diff
@@ -72,20 +73,19 @@ export default function ConsoleLogStream(props: ConsoleLogStreamProps) {
...
+    return Math.min(Math.max(linesHeight, min), max);
   };
 
   return (
     <>
       <Virtuoso
         style={{
           minHeight: `${minHeight()}vh`,
-          maxHeight: window.innerHeight / 2,
+          maxHeight: window.innerHeight * props.maxHeighScale,
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx:88]:
```suggestion
          maxHeight: window.innerHeight * props.maxHeightScale,
```

**Comment by timja** [comment]
I found on my 13" MBP that it didn't display as well as yours on short logs at least.

e.g

```
node {
stage('hi') {
    sh """
        echo 1
    """
    
    sh """
        echo 1
        echo 2
    """
    
    sh """
        echo 1
        echo 2
        echo 3
        echo 4
        echo 5
    """
    
        sh """
        echo 1
        echo 2
        echo 3
        echo 4
        echo 5
        echo 1
        echo 2
        echo 3
        echo 4
        echo 5
        echo 1
        echo 2
        echo 3
        echo 4
        echo 5
    """
}
}
```

the echo 1 echo2 one should display a bit more:
image

any thoughts?

**Comment by stuartrowe** [comment]
That was my concern. There must be a way to calculate the height of each line so the `Virtuoso` can be sized based on that and the number of log lines. I found one example using `forwardRef`, but I couldn't make it work - the ref was never valid in the `useEffect` hook.

**Comment by mikedld** [comment]
Haven't tried it with these changes but long lines may also wrap so you can't just multiply line height by the number. Is there a way of not specifying the minimum height but just letting the browser do that, and only limit the maximum height to something reasonable (e.g. that fits the viewport)?

**Comment by stuartrowe** [comment]
@mikedld thanks, that's a great point about line wrapping. 

> Is there a way of not specifying the minimum height but just letting the browser do that, and only limit the maximum height to something reasonable (e.g. that fits the viewport)?

One would think so, but that's not the case. I found that if the `minHeight` style was passed to the `Virtuoso` component then it would never be taller than that given height. I assume that's why the current implementation is also calculating `minHeight` proportional to the number of lines. I'm trying to use `height` instead after calculating the actual height of each line item multiplied by the number of lines.

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx [hunk 6]
```diff
@@ -20,6 +21,7 @@ export default function ConsoleLogStream(props: ConsoleLogStreamProps) {
   const [moveToBottom, setMoveToBottom] = useState(true);
   const showButtonInterval = useRef<NodeJS.Timeout | null>(null);
   const [showButton, setShowButton] = useState(false);
+  const [consoleLineHeight, setConsoleLineHeight] = useState(1);
```
Comment (stuartrowe) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx:24]: Setting `1` as a default value feels like a bit of a hack, but the component wouldn't render when set to `0` and then the `consoleLineHeightCallback` never gets called.

File: src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx [hunk 7]
```diff
@@ -50,6 +52,10 @@ export default function ConsoleLogStream(props: ConsoleLogStreamProps) {
     }
   }, [moveToBottom]);
 
+  const consoleLineHeightCallback = useCallback((height: number) => {
+    setConsoleLineHeight(height);
```
Comment (stuartrowe) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx:56]: I can actually sum the height of each ConsoleLine `div` through this callback to accommodate wrapped lines.

Comment (stuartrowe) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/ConsoleLogStream.tsx]: Unfortunately that approach causes visual artifacts as each ConsoleLine is rendered and the Virtuoso component noticeably resizes. The current solution is good enough for the majority of cases. The only issues are when there are only a few log lines with very different heights due to text wrapping. Using the max line height in the calculation is the best way to handle this case.

**Review by timja — APPROVED** [review]
Thanks, much better!

**Pull request merged**